### ISSUE TITLE
False positive AM04(ambiguous.column_cout) on known number of recursive CTE syntax

### ISSUE THREAD

**Issue by kang8** [sqlfluff/sqlfluff#5395]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When I tried to use `*` in recursive CTE, SQLFluff told me not to do so, and I used `*` because they were known columns

### Expected Behaviour

Expect to see no errors.

### Observed Behaviour

```
== [test-AM04-ambigouos-column-count.sql] FAIL                                                                                                                                                                                   
L:  17 | P:   3 | AM04 | Query produces an unknown number of result columns.                                                                                                                                                     
                       | [ambiguous.column_count]
All Finished 📜 🎉!
```

### How to reproduce

```sql
with cte_users as (
  select
    id,
    manager_id,
    name
  from users
),

employees as (
  select
    0 as chain_level,
    cte_users.*
  from cte_users

  union all

  select
    employees.chain_level + 1 as chain_level,
    cte_users.*
  from cte_users
  inner join employees on cte_users.manager_id = employees.id
)

select * from employees
```

### Dialect

snowflake

### Version

```bash
$ sqlfluff --version                                                                                                                                                                                                   10:57:26 
sqlfluff, version 2.3.0
```

### Configuration

```
[sqlfluff]
dialect = snowflake
max_line_length = 0
large_file_skip_byte_limit = 0
exclude_rules = LT09, AL01, AL05

[sqlfluff:indentation]
tab_space_size = 2
indent_unit = space

[sqlfluff:rules:aliasing.table]
aliasing = implicit

[sqlfluff:rules:aliasing.column]
aliasing = explicit

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = lower
[sqlfluff:rules:capitalisation.identifiers]
extended_capitalisation_policy = lower
unquoted_identifiers_policy = aliases
[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = lower
[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower
[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = lower

[sqlfluff:rules:references.special_chars]
quoted_identifiers_policy = none
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterbud** [comment]
Hi @kang8  - thanks for opening this issue.

The rule `AM04` is designed to highlight ambiguous column selection, and your example is the anti-pattern it supposed to prevent, so I don't see a false positive situation here.

The above query scan be re-written to a proper CTE without star selector, something like this:

```sql
with cte_users as (
  select
    id,
    manager_id,
    name
  from users
),

employees as (
  -- Anchor Member: Define the starting point
  select
    0 as chain_level,
    id,
    manager_id,
    name
  from cte_users

  union all

  -- Recursive Member: Join and increment
  select
    employees.chain_level + 1,
    cte_users.id,
    cte_users.manager_id,
    cte_users.name
  from cte_users
  inner join employees on cte_users.manager_id = employees.id
)

-- Final Selection: Explicitly list columns
select 
    chain_level, 
    id, 
    manager_id, 
    name 
from employees;
```

Alternatively you can turn off `AM04`.
